applications/upgradeability-by-proxy.md

```diff
@@ -0,0 +1,202 @@
+# Open Grant Proposal
+
+* **Project Name:** ink! Smart Contract Upgradeability
+* **Team Name:** Fractal
+* **Payment Address:** 0xAd0E9D4CF6B8F2E0C12c6620791164681c99A9ca
+
+## Project Overview :page_facing_up:
+
+### Overview
+
+The main goal of this project is to bring smart contract upgradeability to ink.
+
+Smart contract upgradeability is an important mechanism to fix security bugs in
+contracts. It also enables post-deployment iterative development, by offering
+the flexibility of adding new features.
+
+This mechanism should be flexible enough that contract developers may decide if
+they want to be able to upgrade their contracts, and specify under what
+conditions.
+
+Our team is interested in this because it is very likely that our development
+efforts start with smart contracts before moving to a custom Substrate runtime.
+Another reason upgradeability is particularly important to us is that we expect
+to iterate our logic frequently based on a continuous product discovery process.
+
+
+### Project Details
+
+Upgrade functionality can be implemented in user-land, or as part of the smart
+contracts pallet. We are starting with the first option: it gives us more
+flexibility to experiment, and iterate faster towards an adequate soltuion.
+When we're happy with the result, we'll see what can be migrated to ink as a
+directive to reduce boilerplate.
+
+We're going to start by implementing the proxy pattern. There are a few
+concerns that will need to be addressed:
+
+- figure out how to handle the separation between state and code
+- design a standard for data migrations on contract upgrade
+- design a standard for upgrade authorization
+- decide if the contract's public interface can/should be upgradable or static
+
+We want to allow for flexibility when deciding if a smart contract can be
+upgradable or not, and under what circumstances. Should we allow upgrades to
+the authorization code?
+
+Upgrading the contract's public interface may not be possible when using the
+proxy pattern.
+
+Ink's documentation and examples provide us with the delegator pattern. This
+contract works by calling `put_code` once for the delegator contract and once
+for the "delegated-to" contract. Every method that is delegated must be
+explicitly defined in the delegator.
+
+For milestone 1, we will build a template using this pattern with upgrade and
+authorization methods. This is a template and not a library/ink directive
+because we want to focus on the upgrade pattern mechanism and leave any ink
+quality-of-life improvements, like directives or syntax sugar, for the second
+milestone.
+
+For milestone 2, we will extract this pattern into an ink directive which will
+automatically generate the proxied methods. This brings us ergonomy of use,
+making upgradeability easy to opt in to.
```

Comment:
I am not sold on this milestone as it is written.

It would be better if the second milestone depends on the outcomes of the first one.
Having built-in support in ink! for upgradeable smart contracts is not necessarily an improvement to the overall design and should be taken with great care and consideration.
Depending on the outcome of milestone 1 it might be more helpful if generating the base template would be built-in into tooling such as `cargo-contract` instead of the language itself.
Also there might be a chance that implementing a specialized built-in feature would compromise implementation of a more general utility into the language that would solve the problem equally well or better.

Reply:
Yes. I would very much like to see the Milestone 2 be changed to something like: "Merge the template into the official repository and improve cargo-contract to make use of it by generating a proxy upgradeable contract scheme". 

There are just to many uncertainties whether this would make a good language feature or not. I would rather see some real world usage and iterations on the "official standard template" before thinking about a language extensions.

Maybe this could be a follow up grant.